File: src/test/java/com/lsd/parse/SynchronousResponseParserTest.java
package com.lsd.parse;

import com.lsd.IdGenerator;
import com.lsd.events.SynchronousResponse;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

public class SynchronousResponseParserTest {
    private final IdGenerator deterministicIdGenerator = new IdGenerator(true);
    private final Parser parser = new SynchronousResponseParser(deterministicIdGenerator);

    @ParameterizedTest
    @ValueSource(strings = {
            "This is clearly not a sequence event",
            "sync response from A",
            "message from A to B"
    })
    void shouldNotParseInvalidPatterns(String invalidInput) {
        assertThat(parser.parse(invalidInput, null)).isEmpty();
    }

    @ParameterizedTest
    @ValueSource(strings = {
            "sync response from A to B",
            "sync message from A to B [#red]",
            "sync some label from A to B",
            "sync from A to B [#red]"
    })
    void shouldParseValidPatterns(String validInput) {
        assertThat(parser.parse(validInput, null)).isPresent();
    }

    @Test
    void parseSynchronousResponse() {
        var event = parser.parse("sync call from A to B", "");

        assertThat(event).isEqualTo(Optional.of(SynchronousResponse.builder()
                .id("1")
                .label("call")
                .from("A")
                .to("B")
                .colour("")
                .data("")
                .build()));
    }

    @Test
    void parseSynchronousResponseWithColour() {
        var event = parser.parse("sync response message from A to B [#red]", "body");

        assertThat(event).isEqualTo(Optional.of(SynchronousResponse.builder()
                .id("1")
                .label("response message")
                .from("A")
                .to("B")
                .colour("red")
                .data("body")
                .build()));
    }
}

File: src/test/kotlin/com/lsd/parse/SynchronousResponseParserTest.kt
package com.lsd.parse

import com.lsd.IdGenerator
import com.lsd.events.SynchronousResponse
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

class SynchronousResponseParserTest {

    private val deterministicIdGenerator = IdGenerator(true)
    private val parser: Parser = SynchronousResponseParser(deterministicIdGenerator)

    @ParameterizedTest
    @ValueSource(strings = ["This is clearly not a sequence event", "sync response from A", "message from A to B"])
    fun shouldNotParseInvalidPatterns(invalidInput: String) {
        assertThat(parser.parse(invalidInput, "")).isNull()
    }

    @ParameterizedTest
    @ValueSource(strings = ["sync response from A to B", "sync message from A to B [#red]", "sync some label from A to B", "sync from A to B [#red]"])
    fun shouldParseValidPatterns(validInput: String) {
        assertThat(parser.parse(validInput, "")).isNotNull
    }

    @Test
    fun parseSynchronousResponse() {
        val event = parser.parse("sync call from A to B", "")
        assertThat(event).isEqualTo(
            SynchronousResponse.builder()
                .id("1")
                .label("call")
                .from("A")
                .to("B")
                .colour("")
                .data("")
                .build()
        )
    }

    @Test
    fun parseSynchronousResponseWithColour() {
        val event = parser.parse("sync response message from A to B [#red]", "body")
        assertThat(event).isEqualTo(
            SynchronousResponse.builder()
                .id("1")
                .label("response message")
                .from("A")
                .to("B")
                .colour("red")
                .data("body")
                .build()
        )
    }
}